pkg/reconciler/brokercell/brokercell.go

```diff
@@ -42,22 +66,85 @@ var _ bcreconciler.Finalizer = (*Reconciler)(nil)
 func (r *Reconciler) ReconcileKind(ctx context.Context, bc *intv1alpha1.BrokerCell) pkgreconciler.Event {
 	bc.Status.InitializeConditions()
 
+	// Reconcile ingress deployment and service
+	ingressArgs := r.makeIngressArgs(bc)
+	if _, err := reconciler.ReconcileDeployment(r.kubeClientSet, r.deploymentLister, resources.MakeIngressDeployment(ingressArgs)); err != nil {
+		r.Logger.Errorf("Failed to reconcile ingress deployment for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
+		bc.Status.MarkIngressFailed("IngressDeploymentFailed", "Failed to reconcile ingress deployment for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
+		return err
+	}
+	endpoints, err := reconciler.ReconcileService(r.kubeClientSet, r.serviceLister, r.endpointsLister, resources.MakeIngressService(ingressArgs))
+	if err != nil {
+		r.Logger.Errorf("Failed to reconcile ingress service for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
+		bc.Status.MarkIngressFailed("IngressServiceFailed", "Failed to reconcile ingress service for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
+		return err
+	}
+	bc.Status.PropagateIngressAvailability(endpoints)
+	bc.Status.IngressTemplate = "http://" + names.ServiceHostName(endpoints.GetName(), endpoints.GetNamespace())
```

Comment:
The intent here is to signal to the broker controller how to generate the address for a broker, so it doesn't have to embed its knowledge of the ingress. It's kind of a synthetic and maybe useless separation, since the broker controller and brokercell controller are always in the same container.

Since this field isn't used by the broker controller yet, it should probably just be empty (or removed from the struct entirely, but that's a job for another PR).

Reply:
I like the separation though and prefer to keep this :) Although we can use static names for now, but in theory Broker controller doesn't know the ingress endpoint. BrokerCell controller should know since it creates the ingress. 

Is there any strong reason not to keep this?

Reply:
:smiley: No strong reason. If we want to keep this, then it should be a URI template so the Broker controller doesn't have to know that the path is `/namespace/name`. I'll open an issue for this so it doesn't derail this PR.

Reply:
Thanks opening the issue. I guess that issue is about having Broker controller to use the template? I updated this PR to return a URI template string. Let me know if it's good :)

Reply:
Thanks, looks good. I updated the issue title to be more clear that it's about adding this functionality to broker controller.